feat: add filtering via filenames in asset list endpoint

mavidser opened the pull request:
This PR adds API changes to support adding asset overwrite confirmation in CMS.

**Frontend PR:** https://github.com/openedx/studio-frontend/pull/384

**JIRA Ticket:** TNL-10321

## Testing instructions

From `make studio-shell` run: `python -Wd -m pytest --ds=cms.envs.test cms/djangoapps/contentstore/views/tests/test_assets.py::UploadTestCase::test_pre_existing_asset`

mavidser committed:
feat: Add display_name filtering in GET assets/:course

This allows clients to check if a file already exist before
overwriting the asset with new data. See openedx/studio-frontend#384

openedx-webhooks commented:
Thanks for the pull request, @mavidser!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

mphilbrick211 commented:
Hi @mavidser! Regarding the [CLA](https://openedx.atlassian.net/wiki/spaces/COMM/pages/941457737/How+to+start+contributing+to+the+Open+edX+code+base) message above, can you confirm if you're contributing as an individual, or if you're contributing on behalf of an organization?

If you're contributing as an individual, please fill out the [CLA](https://openedx.atlassian.net/wiki/spaces/COMM/pages/941457737/How+to+start+contributing+to+the+Open+edX+code+base) form as requested. If you are contributing on behalf of an organization, please have your manager reach out to oscm@axim.org to be added to your organization's entity agreement.

tecoholic commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,50 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+    """
+    Get a list of filenames that already exist in the course.
+    """
+    if not has_course_author_access(request.user, course_key):
```
@mavidser Can we move this check to the start of `check_asset_pre_upload` function?

tecoholic commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,50 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+        raise PermissionDenied()
+
+    result = []
+    data = json.loads(request.body.decode('utf8'))
```
nit:
```suggestion
    data = json.loads(request.body.decode('utf-8'))
```

tecoholic commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,50 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+            'allowed': False,
+            'reason': _('The following files already exist: {}').format(', '.join(result))
+        }, status=409)
+    else:
```
Nit: The else here is redundant.

tecoholic reviewed (approved):
@mavidser :+1: I have left a couple of small suggestions. Otherwise LGTM

- [x] I tested this: I tested this along with https://github.com/openedx/studio-frontend/pull/384 by uploading files and verifying the errors are generated as expected.
- [x] I read through the code
- [x] I checked for accessibility issues
- [ ] Includes documentation

mavidser commented:
@tecoholic Thanks, I've addressed the requested changes and rebased :)

tecoholic commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,51 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+
+    Returns a JSON response with the following fields:
+    - allowed: boolean, whether the user is allowed to upload an asset
+    - reason: string, the reason why the user is not allowed to upload an asset
```
@mavidser Can you kindly add the `files` to the docstring here?

mavidser replied:
Done

tecoholic commented:
@jristau1984 @cablaa77 Hi this PR along with https://github.com/openedx/studio-frontend/pull/384 should resolve the ticket TNL-10321. Can you kindly review?

mphilbrick211 commented:
Hi @tecoholic! Just flagging that there's still a failing check :)

mavidser commented:
@mphilbrick211 Those seemed to be some transient build errors. Fixed by rebasing :)

jristau1984 commented:
@mphilbrick211 T&L will not be able to review this PR for a while, as we are working on rolling several new features out through this and next month.

ormsbee commented:
@mphilbrick211, @jristau1984: I can get this one.

ormsbee commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,52 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+    for filename in filenames:
+        location = StaticContent.compute_location(course_key, filename)
+        try:
+            contentstore().find(location)
```
Please pass in `as_stream=True` to the `find` call so that it doesn't pull back all the file data for the static asset–some assets can be quite large, and this could cause performance issues.

ormsbee commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,52 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+    """
+    result = []
+    data = json.loads(request.body.decode('utf-8'))
+    filenames = data['filenames']
```
Small nit (optional): This sort of thing can be a little clearer and easier to test if you isolate the request parsing into one place (your `check_asset_pre_upload` view function in this case). Then this function can just take a list of filenames and check them against the contentstore.

That being said, this function is small and simple enough where I'm fine with merging it in its current state.

ormsbee commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,52 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+    - allowed: boolean, whether the user is allowed to upload an asset
+    - reason: string, the reason why the user is not allowed to upload an asset
+    - files: list of strings, the filenames that already exist
+    """
```
It looks like this endpoint is designed to become a place where we can add additional checks other than just name conflict checking. It also looks like it's making decisions about what to allow and what not to allow (with the 409 error). But this endpoint doesn't really make that decision–the person uploading will be allowed to overwrite if they want to. This endpoint exists to better inform their decision to make sure they're overwriting intentionally.

Given that, maybe it would be better to model this as an endpoint that just returns information about a set of filenames? So say we still POST up a bunch of filenames, and it gives us a few bits of basic information based on the metadata we get back in the contentstore `find` call? Something like:

```javascript
{
  "files": {
    "handout.pdf": {
      "size": 20930,
      "hash": "938c2cc0dcc05f2b68c4287040cfcf71"
    },
    "non-existant-handout.pdf": null
  }
}
```

It could then be a more generically named view that could possibly be used in circumstances other just uploads.

Non-existent files return with `null`, and the frontend code can decide what to do with that information–abort, upload it anyway, etc. If we want to add more metadata later to do more clever things, we can–e.g. check the hash to make sure we're not overwriting with the same file (out of scope for this task). But the backend doesn't say whether something is allowed or not, because technically it's all allowed. It just gives the frontend enough info to decide what should be done.

mavidser replied:
This sounds like a rather great idea. I took this a step further and reused the `GET assets/<course_key>`

ormsbee commented on cms/djangoapps/contentstore/asset_storage_handlers.py:
```diff
@@ -88,6 +88,52 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
+    if not has_course_author_access(request.user, course_key):
+        raise PermissionDenied()
+
+    result = _get_pre_existing_filenames(request, course_key)
```
Please set some limit on the number of files that can be queried against at one time by erroring out immediately if the number of files to check exceeds that limit. 100 maybe? It should be high enough that there's no reasonable use case where it would be hit by a human user. But low enough where it won't bring down the entire process with a timeout trying to pull that much data from MongoDB.

mavidser replied:
Have limited the max upload size to 100 on https://github.com/openedx/edx-platform/pull/33493 , which is controlled by the pageSize in the updated PR

ormsbee reviewed (changes requested):
Thank you for the PR! People have wanted this feature for a while. 😄

mavidser commented:
@ormsbee I've updated the PR to be much more simplistic. The suggestion to use asset details made sense, so I've added `display_name` filtering options to the `/assets/:course_key` endpoint, which the frontend uses to figure out filename conflicts. Also, have implemented limiting of parameter size by using pageSize, which the frontend sets to 100.

ormsbee reviewed (approved):
Please squash your commits and give a little more detail in your commit message about why this enhancement to the endpoint is useful, i.e. the use case it serves. I'll merge this in the morning. Thank you for your contribution!

mavidser commented:
@ormsbee Ack, have squashed :+1:

openedx-webhooks commented:
@mavidser 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/33493&entry.1484655318=https://github.com/mavidser&entry.752974735=2023-10-15+21:22&entry.1917517419=2023-10-26+13:32&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
